### PR TITLE
API: allow mixed-datetimes-and-ints in to_datetime, DatetimeIndex

### LINKED ISSUE #49037: API: Series(mixed_datetime_and_ints, dtype="M8[ns]") vs DatetimeIndex(same) mismatch
```python
values = [pd.Timestamp.now(), 1]

pd.Series(values, dtype="M8[ns]")  # <- casts
pd.DatetimeIndex(values, dtype="M8[ns]")  # <- raises
pd.to_datetime(values) # <- raises
```

These should either all cast or all raise.  I'd be OK either way, leaning towards all-raise.

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #49037 (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

cc @jreback this goes a long way towards getting Series/Index constructors to match, but i dont see the reason for the current behavior, do you remember?

### CODE REVIEW

File: pandas/tests/tools/test_to_datetime.py [hunk 1]
```diff
@@ -1444,9 +1444,14 @@ def test_unit_mixed(self, cache, exp, arr):
         result = to_datetime(arr, errors="coerce", cache=cache)
         tm.assert_index_equal(result, expected)
 
-        msg = "mixed datetimes and integers in passed array"
-        with pytest.raises(ValueError, match=msg):
-            to_datetime(arr, errors="raise", cache=cache)
+        # GH#49037 pre-2.0 this raised, but it always worked with Series,
+        #  was never clear why it was disallowed
+        result = to_datetime(arr, errors="raise", cache=cache)
```
Comment (mroeschke) [inline pandas/tests/tools/test_to_datetime.py:1449]: Do we have a test where `unit` and/or `origin` is also specified in this mixed cases?

**Review by mroeschke — COMMENTED** [review]
Generally +1 to have `[ints, datetimes]` be supported, especially in `to_datetime`

**Comment by jorisvandenbossche** [comment]
I commented on the issue (https://github.com/pandas-dev/pandas/issues/49037), I think I am also +1 on allowing the mixture (although I don't like that specific behaviour, it seems the most consistent if we want to have generally applicable rules)

@jbrockmendel one thing I am wondering: the `allow_mixed` keyword, that's not useful to keep for potential stricter behaviour? (although I assume it's not actually useful for "only datetime objects", since `allow_mixed=False` can also mean that you have only ints)

**Comment by jbrockmendel** [comment]
> one thing I am wondering: the allow_mixed keyword, that's not useful to keep for potential stricter behaviour? (although I assume it's not actually useful for "only datetime objects", since allow_mixed=False can also mean that you have only ints)

Assuming you mean in array_to_datetime.  I don't think its particularly useful, but its also not harmful (except for not being tested).  For only-datetime-objects (i.e. the strict from_scalars) i think it may be cleaner to implement a new dedicated function than further overloading array_to_datetime.

File: pandas/tests/tools/test_to_datetime.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/tests/tools/test_to_datetime.py:1439]: i dont think so, no

**Comment by jbrockmendel** [comment]
failures unrelated

File: pandas/tests/tools/test_to_datetime.py (hunk 1 above)
Comment (mroeschke) [reply pandas/tests/tools/test_to_datetime.py:1439]: Could this test be easily added? ( I would expect `unit` and `origin` only applicable to the ints and ignored for datetimes )

Comment (jbrockmendel) [reply pandas/tests/tools/test_to_datetime.py:1439]: origin is only relevant with numeric dtypes, so won't go through array_to_datetime

**Review by mroeschke — APPROVED** [review]
LGTM

**Comment by mroeschke** [comment]
Thanks @jbrockmendel

**Pull request merged**